apps/emqx_management/src/emqx_mgmt_api_publish.erl

```diff
@@ -167,10 +337,23 @@ messages([MessageMap | List], Res) ->
             {error, R}
     end.
 
-format_message_response(Messages) when is_list(Messages) ->
-    [format_message_response(Message) || Message <- Messages];
-format_message_response(#message{id = ID}) ->
-    #{id => emqx_guid:to_hexstr(ID)}.
+make_publish_response(#message{id = ID}) ->
+    #{
+        id => emqx_guid:to_hexstr(ID)
+    }.
+
+make_publish_error_response(ReasonCode) ->
+    make_publish_error_response(ReasonCode, emqx_reason_codes:name(ReasonCode)).
+
+make_publish_error_response(ReasonCode, Msg) ->
+    #{
+        reason_code => ReasonCode,
+        message => to_binary(Msg)
+    }.
 
+to_binary(Atom) when is_atom(Atom) ->
+    atom_to_binary(Atom);
+to_binary(Msg) when is_binary(Msg) ->
+    Msg;
 to_binary(Term) ->
-    list_to_binary(io_lib:format("~p", [Term])).
+    list_to_binary(io_lib:format("~0p", [Term])).
```

Comment:
Q: is this equivalent to `iolist_to_binary`?  I'm wondering if there's some output of `io_lib:format` that is not a flat list :thinking: